libraries/previousAuctionInfo/previousAuctionInfo.js

```diff
@@ -0,0 +1,130 @@
+      });
+    }
+
+    if (auctionDetails.bidsRejected && auctionDetails.bidsRejected.length) {
+      auctionDetails.bidsRejected.forEach(bidRejected => {
+        rejectedBidsMap[bidRejected.requestId] = bidRejected;
+      });
+    }
+
+    if (auctionDetails.bidderRequests && auctionDetails.bidderRequests.length) {
+      auctionDetails.bidderRequests.forEach(bidderRequest => {
+        const enabledBidder = enabledBidders.find(bidder => bidder.bidderCode === bidderRequest.bidderCode);
+
+        if (enabledBidder) {
+          bidderRequest.bids.forEach(bid => {
+            const isValidBid = enabledBidder.isBidRequestValid(bid);
+
+            if (!isValidBid) return;
+
+            const previousAuctionInfoPayload = {
+              bidderRequestId: bidderRequest.bidderRequestId,
+              minBidToWin: highestCpmBid?.cpm || 0,
+              rendered: 0,
+              transactionId: bid.ortb2Imp.ext.tid || bid.transactionId,
+              source: 'pbjs',
+              auctionId: auctionDetails.auctionId,
+              impId: bid.adUnitCode,
+              // bidResponseId: FLOAT, // don't think this is available client side?
+              // targetedbidcpm: FLOAT, // don't think this is available client side?
+              highestcpm: highestCpmBid?.cpm || 0,
+              cur: bid.ortb2.cur,
+              bidderCpm: receivedBidsMap[bid.bidId] ? receivedBidsMap[bid.bidId].cpm : 'nobid',
+              biddererrorcode: rejectedBidsMap[bid.bidId] ? rejectedBidsMap[bid.bidId].rejectionReason : -1,
+              timestamp: auctionDetails.timestamp,
+            }
+
+            window.pbpai = window.pbpai || {};
+            if (!window.pbpai[bidderRequest.bidderCode]) {
+              window.pbpai[bidderRequest.bidderCode] = [];
+            }
+
+            if (window.pbpai[bidderRequest.bidderCode].length > enabledBidder.maxQueueLength) {
+              window.pbpai[bidderRequest.bidderCode].shift();
+            }
+
+            window.pbpai[bidderRequest.bidderCode].push(previousAuctionInfoPayload);
+          });
+        }
+      });
+    }
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(error);
+  }
+}
+
+const onBidWonHandler = (winningBid) => {
+  // // eslint-disable-next-line no-console
+  // console.log('onBidWonHandler', winningBid);
+  winningBidsMap[winningBid.transactionId] = winningBid;
+}
+
+export const onBidRequestedHandler = (bidRequest) => {
+  try {
+    const enabledBidder = enabledBidders.find(bidder => bidder.bidderCode === bidRequest.bidderCode);
+    window.pbpai = window.pbpai || {};
+
+    if (enabledBidder && window.pbpai[bidRequest.bidderCode]) {
+      window.pbpai[bidRequest.bidderCode].forEach(prevAuctPayload => {
+        if (winningBidsMap[prevAuctPayload.transactionId]) {
+          prevAuctPayload.minBidToWin = winningBidsMap[prevAuctPayload.transactionId].cpm;
+          prevAuctPayload.rendered = 1;
+        }
+      });
```

Comment:
The purpose of this is to update the prevAuctionInfo object with the winning bid's cpm, right? If so, I think it'd be more intuitive to put it in the BID_WON handler. But see also below on transactionId